design/plugin-backup-and-restore-progress-design.md

```diff
@@ -0,0 +1,448 @@
+
+Plugin will be able to identify CR by using snapshotID that it received as parameter of CreateVolumeFromSnapshot API, and plugin's Provider name.
+It updates the progress of restore operation regularly if plugin supports feature of showing progress.
+
+Velero deletes VolumePluginRestore CR when it handles deletion of Restore CR.
+
+### Approach 2
+
+This approach is different to approach 1 only with respect to Backup.
+
+#### VolumePluginBackup CR
+
+```
+// VolumePluginBackupSpec is the specification for a VolumePluginBackup CR.
+type VolumePluginBackupSpec struct {
+	// Volume is the PV name to be backed up.
+	Volume string `json:"volume"`
+
+	// Backup name
+	Backup string `json:"backup"`
+
+	// Provider is the Provider field given in VolumeSnapshotLocation
+	Provider string `json:"provider"`
+}
+
+// VolumePluginBackupStatus is the current status of a VolumePluginBackup CR.
+type VolumePluginBackupStatus struct {
+	// Phase is the current state of the VolumePluginBackup.
+	Phase string `json:"phase"`
+
+	// SnapshotID is the identifier for the snapshot of the volume.
+	// This will be used to relate with output in 'velero describe backup'
+	SnapshotID string `json:"snapshotID"`
+
+	// Message is a message about the volume plugin's backup's status.
+	// +optional
+	Message string `json:"message,omitempty"`
+
+	// StartTimestamp records the time a backup was started.
+	// Separate from CreationTimestamp, since that value changes
+	// on restores.
+	// The server's time is used for StartTimestamps
+	// +optional
+	// +nullable
+	StartTimestamp *metav1.Time `json:"startTimestamp,omitempty"`
+
+	// CompletionTimestamp records the time a backup was completed.
+	// Completion time is recorded even on failed backups.
+	// Completion time is recorded before uploading the backup object.
+	// The server's time is used for CompletionTimestamps
+	// +optional
+	// +nullable
+	CompletionTimestamp *metav1.Time `json:"completionTimestamp,omitempty"`
+
+	// PluginSpecific are a map of key-value pairs that plugin want to provide
+	// to user to identify plugin properties related to this backup
+	// +optional
+	PluginSpecific map[string]string `json:"pluginSpecific,omitempty"`
+
+	// Progress holds the total number of bytes of the volume and the current
+	// number of backed up bytes. This can be used to display progress information
+	// about the backup operation.
+	// +optional
+	Progress VolumeOperationProgress `json:"progress,omitempty"`
+}
+
+type VolumeOperationProgress struct {
+	TotalBytes int64
+	BytesDone int64
+}
+
+type VolumePluginBackup struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec VolumePluginBackupSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status VolumePluginBackupStatus `json:"status,omitempty"`
+}
+```
+
+For every backup operation of volume, volume snapshotter creates VolumePluginBackup CR in Velero namespace.
```

Comment:
suggest moving creation of the customresource into the plugin.

Reply:
responded in above comment